### ISSUE TITLE
Plugins

### ISSUE THREAD

**Issue by Lockzi** [jishi/node-sonos-http-api#528]
I would like to add MQTT as an alternative for webhook, but that would cause another dependency (mqtt lib).

What about a plugin architecture?

**Comment by jishi** [comment]
I can see how mqtt could be convenient, but it would almost be equally easy to write a separate service that just publishes all emits from sonos-discovery onto mqtt without sonos-http-api at all. 

Optionally just build a webhook to mqtt bridge. No real point in building one service to rule them all, IMO. The webhook was easy since it only uses core functionality from node.js, as you say, any other transport would probably require additional dependencies that would be unnecessary for the broader masses.

**Comment by Lockzi** [comment]
I agree with everything, except that there are already so many nice things in the http-api (TTS, Presets, all the actions) that if I would fork it and make an node-sonos-mqtt-api it would be hard to keep it up to date with all the nice features being implemented here (not HTTP related).

The best solution may for just this MQTT idea be to write an HTTP<->MQTT bridge, but what about the #516 and then later the #xxx?

That's why I purposed a plugin architecture, but that may end up with that HTTP would be a plugin as well and then this repo would be the "rule them all" you talk about.

I had a look at https://github.com/c9/architect - would it be a good solution even though it would require some refactoring?

**Comment by jishi** [comment]
I'm not saying forking the sonos-http-api, but rather have an auxiliary service which sole purpose is to send mqtt messages. All of the webhook events are actually produced from the sonos-discovery library and is basically just requests from the players that are formatted and forwarded. You would just run both the sonos-http-api and a sonos-mqtt service in parallel. 

Doing it modulary is not really a problem though, the actions today is actually modularized, you can add any javascript fil into the `actions` folder and it would load given that you actually follow the defined API. I just want to avoid putting everything in on app if there is no real need for it.

**Comment by NovaGL** [comment]
I guess the problem with webhook is if that server is not alive how can the MQTT data be pushed from it.

I like MQTT much better has it has things like Last will, which sends then message before it dies if that happens. When new clients join they get the latest message with webhook that's not possible.

You have added extra TTS services (unrelated i know) so maybe the option for MQTT in the settings.js wouldn't be a bad idea.

**Comment by Lockzi** [comment]
I already have a version running where i use the web hook setting but with mqtt:// and intercept the webbhook, but like my initial suggestion it now forces a dependency of MQTT unless we get momentum for a plugin architecture.

**Comment by NovaGL** [comment]
Can you share how you did that. I would be interested

**Comment by Lockzi** [comment]
Updated the response above as I wrote it initially on mobile.

Not hard at all, just implement an MQTT connection and create an equal invokeWebhook() function. It is however not subscibing for commands yet, only publishing.

**Comment by NovaGL** [comment]
Thanks I'm only interested in subscribing

**Comment by jishi** [comment]
I'm not following your logic:

> I've had a look on how to build a bridge as well but it seems wrong to go from MQTT, to an HTTP request, to Sonos. In my head the reversed would play out far better.

> The only way I see to get propper MQTT support is by having it in the same code to access all the functionality directly. Having a different application is just generating double workload (MQTT sends an HTTP request to HTTP-api and HTTP api executes it). Further more, how often would you refresh /states for every player? (I have 11 sonos units)

Now you are talking about controlling Sonos via MQTT, I thought you we're talking about notifications (subscribing) because you we're talking about alternative for webhook?

And why would you refresh /states? The webhook is invoked when state changes (except for elapsed time). That trigger would also create an mqtt message, in my head :)

**Comment by Lockzi** [comment]
The focus is the same: plugin architecture. The reasons why has somewhat shifted ;)
It's so convinient if you already am publishing, to have the urge to subscribe. 

What I have now is a solution that is only able to publish, just like webhooks, but that solution is not feasible to push without it being rendered as another another branch due to further dependencies. If that was to be done, the project might as well be called node-sonos-mqtt-api as the subscribing feature would quickly take phase, and probably by that branches users replace the HTTP usage (so why would node-sonos-mqtt branch have HTTP parts in it?).

Maybe it's only me and NovaGL who is interested in adding/replacing MQTT (let's say publishing/subscribing/last wish) support, but I am sure if an plugin architecture was in place other possibilities would emerge here. Exciting times 👍

**Comment by jishi** [comment]
I think there might be some interest in mqtt, but as you say, it's probably gonna be either/or, hence, might not make sense to mix the two.

From my point of view, it might make more sense to move the http-api specific features (tts, clips, etc) to the sonos-discovery, make it transport agnostic and then make another wrapper for mqtt. 

If you need both mqtt and http (for whatever reason), then you just run them both. 

What do you think about that?

**Comment by Lockzi** [comment]
That would work excellent for this purpose! :) 👍

I also suspect that that would be a "minor" fix compared to plugin architecture?

**Comment by NovaGL** [comment]
I don't like the idea as I am only using mqtt for subscribing I prefer http for performing the action.

I just want a webhook or MQTT option for event subscription

**Comment by jishi** [comment]
@NovaGL You think it's a problem to run two separate processes, one for mqtt and one for http then? Is it because of memory constraints, or just because it seems unnecessary? It's not a huge overhead to run a second node process, but for machines with very little memory (<128MB) it might of course be an issue.

**Comment by NovaGL** [comment]
I just think its not required. (Or because I find HTTP simpler)

If possible have the option for both.
In an ideal world I would like the events to go to the topic
/sonos-api/{room-name}

Rather than having a generic one called /sonos-api/events.

Look at this project that I use that offers both, it seems to have a low overhead and I run it side by side with Sonos-http-api and MQTT server with no degradation 
https://github.com/maddox/harmony-api

**Comment by NovaGL** [comment]
What I have done is I get the webhook.
Retrieve data.roomName and type and store in in MQTT

So assuming the room name is bedroom and the type it transport-state. 
It would be stored like this
/sonos-api/bedroom/transport-state

 I have to try it out and see if it works but i think it might be a good solution

**Comment by Lockzi** [comment]
@jishi I hope you're enjoying the Swedish summer so far!

Where do you feel we are at regarding this last suggestion before NovaGLs personal solution?

Do you agree with moving the etc. parts (TTS, clips and so forth) to sonos-discovery to allow for http-api to be just concerned with HTTP request and allow another MQTT-api, and if so, how do we proceed?

**Comment by jishi** [comment]
@Lockzi I think that would be the best way forward. I was concerned with how the sonos-discovery was suposed to serve the TTS or clip files to the Sonos system, given that it wouldn't have the same HTTP server, but the discovery can spin up another HTTP server for that or use the same one that listens to events, and just require configuration regarding the location on disk. 

I'll see if I can get some time around looking into this in the near future.

**Comment by besynnerlig** [comment]
> From my point of view, it might make more sense to move the http-api specific features (tts, clips, etc) to the sonos-discovery, make it transport agnostic and then make another wrapper for mqtt.

I'd love to see that happen.

Today I use node-sonos-http-api side by side with "another node.js sonos/MQTT bridge". The latter is not as feature rich as node-sonos-http-api. I'd prefer to use only node-sonos-http-api if it had MQTT support. MQTT is definitely the way to go for bridging with home automation systems.
